File: projects/tools/tools/rotate_device_tokens.lua
-- Rotate device tokens from OLD_WEB_API_SECRET to NEW_WEB_API_SECRET
local redis = require('redis')
local json = require('libs.JSON')
local os = require('os')

local OLD = os.getenv('OLD_WEB_API_SECRET') or os.getenv('WEB_API_SECRET') or ''
local NEW = os.getenv('NEW_WEB_API_SECRET') or ''
if NEW == '' then
  print('Set NEW_WEB_API_SECRET to the new secret (env)')
  os.exit(1)
end

local conn = redis.connect('127.0.0.1', 6379)

local function shell_escape_single(s)
  if not s then return '' end
  return s:gsub("'", "'\\''")
end

local function decrypt_with_secret(cipher, secret)
  if secret == '' or not cipher then return nil, 'no-secret' end
  local c = shell_escape_single(cipher)
  local sec = shell_escape_single(secret)
  local cmd = "printf '%s' '" .. c .. "' | openssl enc -d -aes-256-cbc -a -pass pass:'" .. sec .. "' 2>/dev/null"
  local fh = io.popen(cmd, 'r')
  if not fh then return nil, 'no-openssl' end
  local out = fh:read('*a')
  fh:close()
  if not out or out == '' then return nil, 'decrypt-failed' end
  out = out:gsub('%s+$','')
  return out, nil
end

local function encrypt_with_secret(plain, secret)
  if secret == '' or not plain then return plain end
  local p = shell_escape_single(plain)
  local sec = shell_escape_single(secret)
  local cmd = "printf '%s' '" .. p .. "' | openssl enc -aes-256-cbc -a -salt -pbkdf2 -pass pass:'" .. sec .. "'"
  local fh = io.popen(cmd, 'r')
  if not fh then return nil, 'no-openssl' end
  local out = fh:read('*a')
  fh:close()
  if not out or out == '' then return nil, 'encrypt-failed' end
  out = out:gsub('%s+$','')
  return out, nil
end

local items = conn:lrange('web:devices', 0, -1) or {}
print('Devices to process:', #items)
local new_list = {}
for i,v in ipairs(items) do
  local ok, obj = pcall(function() return json:decode(v) end)
  if not ok or not obj then
    print('Skipping invalid JSON at index', i)
  else
    local token = obj.token
    local plain = nil
    if token then
      local dec, derr = decrypt_with_secret(token, OLD)
      if not dec then
        -- maybe token is in plaintext already
        print('Warning: cannot decrypt token for', obj.id, 'assuming plaintext')
        dec = token
      end
      local enc, err = encrypt_with_secret(dec, NEW)
      if not enc then print('Failed to encrypt for', obj.id, err); enc = dec end
      obj.token = enc
    end
    table.insert(new_list, json:encode(obj))
  end
end

-- replace list atomically
conn:del('web:devices')
for i,v in ipairs(new_list) do conn:rpush('web:devices', v) end
print('Rotation complete. Replaced', #new_list, 'devices')

File: projects/tools/tools/rotate_device_tokens.py
#!/usr/bin/env python3
"""
Rotate device tokens from OLD_WEB_API_SECRET to NEW_WEB_API_SECRET.
Uses OpenSSL on PATH for encryption/decryption to preserve compatibility with existing data.
"""
import os
import sys
import json
import shutil
import subprocess

try:
    import redis
except Exception as e:
    print('Missing dependency: redis', e)
    sys.exit(1)


def has_openssl():
    return shutil.which('openssl') is not None


def decrypt_with_secret(cipher: str, secret: str):
    if not secret or not cipher:
        return None, 'no-secret-or-cipher'
    if not has_openssl():
        return None, 'no-openssl'
    try:
        p = subprocess.run(
            ['openssl', 'enc', '-d', '-aes-256-cbc', '-a', '-pass', f'pass:{secret}'],
            input=cipher.encode('utf-8'),
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            check=False,
            timeout=10,
        )
        out = p.stdout.decode('utf-8')
        out = out.rstrip('\n\r')
        if out == '':
            return None, 'decrypt-failed'
        return out, None
    except Exception as e:
        return None, f'decrypt-exc:{e}'


def encrypt_with_secret(plain: str, secret: str):
    if not secret or plain is None:
        return plain, 'no-secret-or-plain'
    if not has_openssl():
        return None, 'no-openssl'
    try:
        p = subprocess.run(
            ['openssl', 'enc', '-aes-256-cbc', '-a', '-salt', '-pbkdf2', '-pass', f'pass:{secret}'],
            input=plain.encode('utf-8'),
            stdout=subprocess.PIPE,
            stderr=subprocess.DEVNULL,
            check=False,
            timeout=10,
        )
        out = p.stdout.decode('utf-8')
        out = out.rstrip('\n\r')
        if out == '':
            return None, 'encrypt-failed'
        return out, None
    except Exception as e:
        return None, f'encrypt-exc:{e}'


def main():
    OLD = os.getenv('OLD_WEB_API_SECRET') or os.getenv('WEB_API_SECRET') or ''
    NEW = os.getenv('NEW_WEB_API_SECRET') or ''
    if NEW == '':
        print('Set NEW_WEB_API_SECRET to the new secret (env)')
        sys.exit(1)

    REDIS_URL = os.getenv('REDIS_URL', 'redis://127.0.0.1:6379')
    try:
        r = redis.from_url(REDIS_URL, decode_responses=True)
    except Exception as e:
        print('Redis connection error:', e)
        sys.exit(1)

    try:
        items = r.lrange('web:devices', 0, -1) or []
    except Exception as e:
        print('Redis error reading web:devices:', e)
        sys.exit(1)

    print('Devices to process:', len(items))
    new_list = []
    for i, v in enumerate(items, start=1):
        try:
            obj = json.loads(v)
        except Exception:
            print('Skipping invalid JSON at index', i)
            continue
        token = obj.get('token')
        if token:
            dec, derr = decrypt_with_secret(token, OLD)
            if not dec:
                print('Warning: cannot decrypt token for', obj.get('id', f'index {i}'), 'assuming plaintext (err:', derr, ')')
                dec = token
            enc, err = encrypt_with_secret(dec, NEW)
            if not enc:
                print('Failed to encrypt for', obj.get('id', f'index {i}'), err)
                enc = dec
            obj['token'] = enc
        new_list.append(json.dumps(obj, separators=(',', ':')))

    # replace list atomically
    try:
        r.delete('web:devices')
        if new_list:
            r.rpush('web:devices', *new_list)
    except Exception as e:
        print('Redis write error:', e)
        sys.exit(1)

    print('Rotation complete. Replaced', len(new_list), 'devices')


if __name__ == '__main__':
    main()
